src/Admin.cpp

```diff
@@ -100,23 +143,118 @@ namespace geopm
+                             string_join(overlap, "\", \"") + "\"\n",
+                             EINVAL, __FILE__, __LINE__);
+        }
+        // Combine settings
+        std::map<std::string, std::string> config_map(override_map);
+        config_map.insert(default_map.begin(), default_map.end());
+        // Check configuration and print it out
+        std::vector<std::string> policy_names;
+        std::vector<double> policy_vals;
+        check_config(config_map, policy_names, policy_vals);
+        return print_config(config_map, override_map, policy_names, policy_vals);
+    }
+
+    void Admin::check_config(const std::map<std::string, std::string> &config_map,
+                             std::vector<std::string> &policy_names,
+                             std::vector<double> &policy_vals)
+    {
+        if (!config_map.size()) {
+            throw Exception("Admin::check_node(): Configuration files do not exist or are empty",
+                            ENOENT, __FILE__, __LINE__);
+        }
+        // Determine if there is an agent set and check the policy
+        auto agent_it = config_map.find("GEOPM_AGENT");
+        auto policy_it = config_map.find("GEOPM_POLICY");
+        if (agent_it != config_map.end()) {
+            std::string agent_name = agent_it->second;
+            policy_names = Agent::policy_names(agent_name);
+            if (policy_it != config_map.end()) {
+                FilePolicy file_policy(policy_it->second, policy_names);
+                policy_vals = file_policy.get_policy();
+            }
+            else {
+                policy_vals.resize(policy_names.size(), NAN);
+            }
+            std::unique_ptr<Agent> agent = agent_factory().make_plugin(agent_name);
+            agent->validate_policy(policy_vals);
+        }
+        else if (policy_it != config_map.end()) {
+            throw Exception("Admin::check_node(): A policy was specified, but not an agent",
+                            EINVAL, __FILE__, __LINE__);
+        }
+    }
+
+    std::string Admin::print_config(const std::map<std::string, std::string> &config_map,
+                                    const std::map<std::string, std::string> &override_map,
+                                    const std::vector<std::string> &policy_names,
+                                    const std::vector<double> &policy_vals)
+    {
+        std::ostringstream result;
+        result << "GEOPM CONFIGURATION\n"
+               << "===================\n\n";
+        for (const auto &config: config_map) {
+            result << "    " << config.first << "=" << config.second;
+            if (override_map.find(config.first) != override_map.end()) {
+                result << " (override)\n";
+            }
+            else {
+                result << " (default)\n";
+            }
+        }
+        if (policy_vals.size()) {
+            result << "\n";
+            result << "AGENT POLICY\n"
```

Comment:
Might want to revisit this with the Policy class refactoring.  If the original JSON file has gaps in the policy values, they get filled in with NAN, so this may have more keys listed than are in the original file.

Reply:
Okay I was worried about what changes from the Policy class I might be needing here.